// @deck.gl/carto/dist/es5/config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildMapsUrlFromBase = buildMapsUrlFromBase;
exports.defaultClassicCredentials = void 0;
exports.getDefaultCredentials = getDefaultCredentials;
exports.setDefaultCredentials = setDefaultCredentials;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _mapsApiCommon = require("./api/maps-api-common");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var defaultClassicCredentials = {
  username: 'public',
  apiKey: 'default_public',
  region: 'us',
  mapsUrl: null
};
exports.defaultClassicCredentials = defaultClassicCredentials;
var defaultCloudNativeCredentials = {
  accessToken: null,
  apiBaseUrl: null,
  mapsUrl: null
};
var credentials = {};
setDefaultCredentials({});

function setDefaultCredentials(opts) {
  var apiVersion = opts.apiVersion || _mapsApiCommon.API_VERSIONS.V2;

  switch (apiVersion) {
    case _mapsApiCommon.API_VERSIONS.V1:
      opts.mapsUrl = opts.mapsUrl || _mapsApiCommon.DEFAULT_MAPS_URL_FORMAT[apiVersion];
      credentials = _objectSpread(_objectSpread({
        apiVersion: apiVersion
      }, defaultClassicCredentials), opts);
      break;

    case _mapsApiCommon.API_VERSIONS.V2:
      opts.mapsUrl = opts.mapsUrl || _mapsApiCommon.DEFAULT_MAPS_URL_FORMAT[apiVersion];
      credentials = _objectSpread(_objectSpread({
        apiVersion: apiVersion
      }, defaultClassicCredentials), opts);
      break;

    case _mapsApiCommon.API_VERSIONS.V3:
      if (!opts.apiBaseUrl) {
        throw new Error("API version ".concat(_mapsApiCommon.API_VERSIONS.V3, " requires to define apiBaseUrl at credentials. Go to https://app.carto.com to get your apiBaseUrl."));
      }

      var apiBaseUrl = opts.apiBaseUrl || defaultCloudNativeCredentials.apiBaseUrl;
      opts.mapsUrl = opts.mapsUrl || buildMapsUrlFromBase(apiBaseUrl);
      credentials = _objectSpread(_objectSpread({
        apiVersion: apiVersion
      }, defaultCloudNativeCredentials), opts);
      break;

    default:
      throw new Error("Invalid API version ".concat(apiVersion, ". Use API_VERSIONS enum."));
  }
}

function getDefaultCredentials() {
  return credentials;
}

function buildMapsUrlFromBase(apiBaseUrl) {
  var suffix = '/v3/maps';

  if (apiBaseUrl.endsWith('/')) {
    suffix = suffix.substring(1);
  }

  return "".concat(apiBaseUrl).concat(suffix);
}
