// @deck.gl/google-maps/dist/es5/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDeckInstance = createDeckInstance;
exports.destroyDeckInstance = destroyDeckInstance;
exports.getViewPropsFromCoordinateTransformer = getViewPropsFromCoordinateTransformer;
exports.getViewPropsFromOverlay = getViewPropsFromOverlay;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _core = require("@deck.gl/core");

var _math = require("math.gl");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var MAX_LATITUDE = 85.05113;

function createDeckInstance(map, overlay, deck, props) {
  if (deck) {
    if (deck.props.userData._googleMap === map) {
      return deck;
    }

    destroyDeckInstance(deck);
  }

  var eventListeners = {
    click: null,
    dblclick: null,
    mousemove: null,
    mouseout: null
  };
  deck = new _core.Deck(_objectSpread(_objectSpread({}, props), {}, {
    style: null,
    parent: getContainer(overlay, props.style),
    initialViewState: {
      longitude: 0,
      latitude: 0,
      zoom: 1
    },
    controller: false,
    userData: {
      _googleMap: map,
      _eventListeners: eventListeners
    }
  }));

  var _loop = function _loop(eventType) {
    eventListeners[eventType] = map.addListener(eventType, function (evt) {
      return handleMouseEvent(deck, eventType, evt);
    });
  };

  for (var eventType in eventListeners) {
    _loop(eventType);
  }

  return deck;
}

function getContainer(overlay, style) {
  var container = document.createElement('div');
  container.style.position = 'absolute';
  Object.assign(container.style, style);

  if (overlay.getPanes) {
    overlay.getPanes().overlayLayer.appendChild(container);
  } else {
    overlay.getMap().getDiv().appendChild(container);
  }

  return container;
}

function destroyDeckInstance(deck) {
  var eventListeners = deck.props.userData._eventListeners;

  for (var eventType in eventListeners) {
    eventListeners[eventType].remove();
  }

  deck.finalize();
}

function getViewPropsFromOverlay(map, overlay) {
  var _getMapSize = getMapSize(map),
      width = _getMapSize.width,
      height = _getMapSize.height;

  var projection = overlay.getProjection();
  var bounds = map.getBounds();
  var ne = bounds.getNorthEast();
  var sw = bounds.getSouthWest();
  var topRight = projection.fromLatLngToDivPixel(ne);
  var bottomLeft = projection.fromLatLngToDivPixel(sw);
  var nwContainerPx = new google.maps.Point(0, 0);
  var nw = projection.fromContainerPixelToLatLng(nwContainerPx);
  var nwDivPx = projection.fromLatLngToDivPixel(nw);
  var leftOffset = nwDivPx.x;
  var topOffset = nwDivPx.y;
  var mapWidth = projection.getWorldWidth();
  var mapCount = Math.ceil(width / mapWidth);
  leftOffset -= Math.floor(mapCount / 2) * mapWidth;
  var scale = height ? (bottomLeft.y - topRight.y) / height : 1;
  var zoom = Math.log2(scale || 1) + map.getZoom() - 1;
  var centerPx = new google.maps.Point(width / 2, height / 2);
  var centerContainer = projection.fromContainerPixelToLatLng(centerPx);
  var latitude = centerContainer.lat();
  var longitude = centerContainer.lng();

  if (Math.abs(latitude) > MAX_LATITUDE) {
    latitude = latitude > 0 ? MAX_LATITUDE : -MAX_LATITUDE;
    var center = new google.maps.LatLng(latitude, longitude);
    centerPx = projection.fromLatLngToContainerPixel(center);
    topOffset += centerPx.y - height / 2;
  }

  return {
    width: width,
    height: height,
    left: leftOffset,
    top: topOffset,
    zoom: zoom,
    pitch: map.getTilt(),
    latitude: latitude,
    longitude: longitude
  };
}

function getViewPropsFromCoordinateTransformer(map, transformer) {
  var _getMapSize2 = getMapSize(map),
      width = _getMapSize2.width,
      height = _getMapSize2.height;

  var _transformer$getCamer = transformer.getCameraParams(),
      center = _transformer$getCamer.center,
      bearing = _transformer$getCamer.heading,
      pitch = _transformer$getCamer.tilt,
      zoom = _transformer$getCamer.zoom;

  var fovy = 25;
  var aspect = height ? width / height : 1;
  var near = 0.75;
  var far = 300000000000000;
  var projectionMatrix = new _math.Matrix4().perspective({
    fovy: fovy * Math.PI / 180,
    aspect: aspect,
    near: near,
    far: far
  });
  var focalDistance = 0.5 * projectionMatrix[5];
  return {
    width: false,
    height: false,
    viewState: {
      altitude: focalDistance,
      bearing: bearing,
      latitude: center.lat(),
      longitude: center.lng(),
      pitch: pitch,
      projectionMatrix: projectionMatrix,
      repeat: true,
      zoom: zoom - 1
    }
  };
}

function getMapSize(map) {
  var container = map.getDiv().firstChild;
  return {
    width: container.offsetWidth,
    height: container.offsetHeight
  };
}

function getEventPixel(event, deck) {
  if (event.pixel) {
    return event.pixel;
  }

  var point = deck.getViewports()[0].project([event.latLng.lng(), event.latLng.lat()]);
  return {
    x: point[0],
    y: point[1]
  };
}

function handleMouseEvent(deck, type, event) {
  var mockEvent = {
    type: type,
    offsetCenter: getEventPixel(event, deck),
    srcEvent: event
  };

  switch (type) {
    case 'click':
      deck._lastPointerDownInfo = deck.pickObject(_objectSpread(_objectSpread({}, mockEvent.offsetCenter), {}, {
        radius: deck.props.pickingRadius
      }));
      mockEvent.tapCount = 1;

      deck._onEvent(mockEvent);

      break;

    case 'dblclick':
      mockEvent.type = 'click';
      mockEvent.tapCount = 2;

      deck._onEvent(mockEvent);

      break;

    case 'mousemove':
      mockEvent.type = 'pointermove';

      deck._onPointerMove(mockEvent);

      break;

    case 'mouseout':
      mockEvent.type = 'pointerleave';

      deck._onPointerMove(mockEvent);

      break;

    default:
      return;
  }
}
